[우아한9조대] 2주차 Facebook 기능 구현

vsh123 opened the pull request:
안녕하세요!
우아한 구조대 2주차 리뷰 요청 드립니다.

이번주 저희가 구현한 기능은 다음과 같습니다.

1. 유저 검색 기능
2. 친구추가, 조회 기능
3. 프로필 페이지 마크업
4. 게시글 좋아요 기능
5. 이미지 업로드 기능

이번 미션을 진행하면서 엔티티 관계 매핑이나, 테이블 구성에 대한 고민이 많았습니다. 코드 보신 후 엔티티 구조에 대한 피드백도 함께 주시면 정말 감사하겠습니다!

그리고 지난주 프론트 구현이 늦어지면서 준의 피드백을 받지 못했습니다. 프론트 부분도 많은 피드백 주시면 감사하겠습니다!

vsh123 committed:
feat: Exception Handler (#43)

choihz committed:
refactor: uri 수정 (#46)

* fix: article api 관련 url에 api 추가

* fix: article, comment 관련 테스트 네이밍 통일

* refactor: comment 관련 url에서 article 관련 내용 제거

vsh123 committed:
feat: 좋아요, 댓글 개수를 알려주는 api 구현 (#44)

* feat: Article Like 개수 구현

* feat: Comment Like 개수 구현

* feat: Article에 따른 Comment 개수 출력 api 구현

* refactor : 컨벤션

* rename: CommentService.getCommentsCountOf()

* refactot: 컨벤션

* fix: uri변경에 따른 테스트코드 수정

vsh123 committed:
feat: 게시글, 댓글 조회 기능 (#45)

* refactor: articleId에 따른 댓글 find 테스트코드 수정

* feat: Article findAll 기능 구현

* refactor: CommentRepository findAll Orederby 추가

* fix: 테스트 코드 수정

choihz committed:
refactor: Article, Comment refactoring (#49)

* feat: 게시글, 댓글 조회 기능 (#45)

* refactor: articleId에 따른 댓글 find 테스트코드 수정

* feat: Article findAll 기능 구현

* refactor: CommentRepository findAll Orederby 추가

* fix: 테스트 코드 수정

* refactor: ResponseEntity 리턴시 static factory method 사용

* feat: ArticleLike에 대하여 ArticleLikeResponse 대신 HttpStatus로 응답

* feat: Article, ArticleLike에 equals, hashCode 추가

* test: ArticleApiController 좋아요 기능 관련 테스트 추가

* refactor: Optimize imports

* refactor: ArticleServiceTest 중복 로직 제거

* refactor: Article의 present 관련 용어를 전부 deleted로 수정

* refactor: Comment의 present 관련 용어를 전부 deleted로 수정

* refactor: CommentServiceTest 중복 로직 제거

* refactor: 사용하지 않는 필드 제거

men7627 committed:
fix: 회원가입 후 자동로그인 구현 (#50)

1. POST("/users")에 login로직, session set로직을 추가한다.
2. signup후 login하는 테스트를 signup만 하게 수정한다.

choihz committed:
feat: ArticleResponse에 recentDate 필드 추가 (#51)

* feat: ArticleResponse에 recentDate 필드 추가

* refactor: getRecentDate에서 null 체크 대신 Optional 사용

choihz committed:
feat: 글쓰기 파일 첨부 모달 및 글쓰기 버튼 구현 (#53)

choihz committed:
refactor: 전체적인 리팩토링 (#54)

* fix: 잘못된 uri (/login) 수정

* refactor: Article isNotAuthor() 구현

* refactor: method 순서 일관성 있게 변경

* refactor: 불필요한 주석 제거

* refactor: 불필요한 공백 제거, 코드 일관성 유지

codeanddonuts committed:
feat: 프로필 페이지 마크업 구현 (#56)

* refactor: app.js

* feat: profile.html

* feat: file upload

men7627 committed:
feat: 자동완성 유저검색 기능 구현 (#55)

* fix: 회원가입 후 자동로그인 구현

1. POST("/users")에 login로직, session set로직을 추가한다.
2. signup후 login하는 테스트를 signup만 하게 수정한다.

* feat: 자동완성 유저검색 기능 구현

1. ajax 통신을 활용해 검색창 datalist를 동적으로 변하게 구현한다.
 - 한글자 입력할 때마다 api요청
2. GET("api/users/{keyword}")를 구현한다.
3. UserRepository에 findByNameContaining(String) 메서드를 추가한다
4. UserService에 findUserNamesByKeyword(String) 메서드를 추가한다.
5. 각 테스트를 추가한다.

* feat: 유저이름 조회 테스트 간소화

vsh123 committed:
feat: 사진 업로드 기능 구현 (#62)

* setting: logback 설정

* feat: ArticleRequest MultipartFile 추가

* feat: Article Multipart 엔티티 생성

* feat: handlebar 이미지 추가

* feat: Article 이미지 첨부 기능 구현

1. 현재 post "/api/artcles/upload"로 구현
 추후 /api/articles와 합쳐야함

* setting: 정적파일 자동 리로드 설정 추가

* feat: 글쓰기(이미지 포함) 구현

* feat: 이미지 로드 기능 구현

* refactor: articleService.saveMultipart()

* refactor: 불필요 요소 제거

men7627 committed:
refactor: 유저 검색 스크롤바 추가 (#63)

* fix: 회원가입 후 자동로그인 구현

1. POST("/users")에 login로직, session set로직을 추가한다.
2. signup후 login하는 테스트를 signup만 하게 수정한다.

* feat: 자동완성 유저검색 기능 구현

1. ajax 통신을 활용해 검색창 datalist를 동적으로 변하게 구현한다.
 - 한글자 입력할 때마다 api요청
2. GET("api/users/{keyword}")를 구현한다.
3. UserRepository에 findByNameContaining(String) 메서드를 추가한다
4. UserService에 findUserNamesByKeyword(String) 메서드를 추가한다.
5. 각 테스트를 추가한다.

* feat: 유저이름 조회 테스트 간소화

* feat: 유저이름 검색 스크롤 추가

eunsukko committed:
feat: 친구추가, 조회기능 추가 (#60)

* fix : typo at application.yml

* feat: id들을 통한 유저들 조회

- 해당 id들을 가지는 유저들을 얻어냄
- 학습테스트 적용

* feat: 친구기능 추가(서비스까지 적용)

- Friendship Entity 추가
	- 두 유저간의 관계를 나타냄
	- 중복되는 값이 없도록 처리 (precedentUserId < userId)
- 간단한 인수 테스트

* fix: User 삭제시 연관된 친구정보 삭제

* feat: 친구요청 api 구현

choihz committed:
feat: Article 좋아요 구현 (#64)

choihz committed:
feat: Article 불러오기 구현 (#65)

* feat: Article 좋아요 구현

* feat: Article 불러오기 구현

* fix: 뉴스피드 출력 관련 오류 수정 및 주석 제거

vsh123 committed:
feat: 친구 추가 버튼 구현 (#66)

* refactor: app.js

* feat: profile.html

* feat: file upload

* feat: profile.html friend add button

* refactor: 컨벤션

phs1116 commented on src/main/java/techcourse/fakebook/controller/ArticleApiController.java:
```diff
@@ -1,51 +1,72 @@
+    public ResponseEntity<ArticleResponse> create(ArticleRequest articleRequest, @SessionUser UserOutline userOutline) {
         ArticleResponse articleResponse = articleService.save(articleRequest, userOutline);
-        return new ResponseEntity<>(articleResponse, HttpStatus.CREATED);
+        return ResponseEntity.created(URI.create("/api/articles")).body(articleResponse);
```
created에서 Location 헤더를 통해서 제공하는 URL은 생성된 자원을 참조할수 있는 완전한 정보입니다. (id값 포함)

phs1116 commented on src/main/java/techcourse/fakebook/controller/ArticleApiController.java:
```diff
@@ -1,51 +1,72 @@
-    public ResponseEntity<ArticleLikeResponse> checkLike(@PathVariable Long id, @SessionUser UserOutline userOutline) {
-        ArticleLikeResponse articleLikeResponse = articleService.isLiked(id, userOutline);
-        return new ResponseEntity<>(articleLikeResponse, HttpStatus.OK);
+    public ResponseEntity<HttpStatus> checkLike(@PathVariable Long id, @SessionUser UserOutline userOutline) {
```
ResponseEntity의 제네릭은 body의 타입입니다.
body가 없으니 Void로 돌려주시면 될 것 같아요. :)

phs1116 commented on src/main/java/techcourse/fakebook/domain/friendship/Friendship.java:
```diff
@@ -0,0 +1,84 @@
+    public Friendship() {
+    }
+
+    public Friendship(User precedentUser, User user) {
```
precedentUser가 항상 앞서있어야한다면 이 생성자는 private로 막고 Friendship.from으로만 강제하는게 낫지 않을까요?

woowahan-pjs replied:
동의합니다. 👍

eunsukko replied:
private 로 작성하였습니다! (아직 `@Entity`가 붙은 애들의 경우 어떤식으로 접근자를 놓아야하는지, getter 와 setter는 정말 필요한 것인지 등이 이해가 되지 않았네요)

기본 생성자가 존재만 하면 (프록시?를 위해선 default 접근자만 있어도) 객체를 잘 생성한다고 이해를 했습니다.

궁금한 부분은 (아직 문서에서 해당 내용을 잘 못찾겠네요..ㅠ) 
setter 가 없어도 db에서 잘 읽어오고 있습니다.
이 경우 생성자를 통해서 넣어주는 걸지.. 아니면 그냥 필드에 넣어주는 건지 혼란이 왔습니다..
(리플랙션? 을 통해서 생성하는 것 같은데.. 리플랙션을 몰라서 어떤 식으로 값을 채워주는지 짐작조차 안가네요...)
문서에서는 어떻게 이것이 가능한지보단.. 자바빈 규약?에 따라서 getter, setter를 놓는 것을 추천하는 것 같은데...

이럴 경우... 어떻게 가능한지 예측하기보단.. 그냥 문서에 있는 것을 따르는 것이 좋을까요??
(지금은 setter를 최대한 피하고 싶은 마음에 setter를 빼놓았었습니다. 그래서.. 현재는 'setter를 빼도, 생성자를 private 로 해도 db에서 해당 엔터티를 읽어올 수 있구나'를 확인한 정도입니다. )

phs1116 replied:
현재 하이버네이트를 사용하고 있으므로 하이버네이트 쪽 문서를 참고하시면 될 것 같습니다.
물론 하이버네이트 외에 다른 구현체라면 private 생성자를 통해서 생성을 지원하지 않을 수 있습니다.

https://docs.jboss.org/hibernate/orm/5.4/quickstart/html_single/#hibernate-gsg-tutorial-basic-entity

https://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#entity

- This class uses standard JavaBean naming conventions for property getter and setter methods, as well as private visibility for the fields. Although this is the recommended design, it is not required.

- The no-argument constructor, which is also a JavaBean convention, is a requirement for all persistent classes. Hibernate needs to create objects for you, using Java Reflection. The constructor can be private. However, package or public visibility is required for runtime proxy generation and efficient data retrieval without bytecode instrumentation.

- The persistent state of an entity is represented by instance variables, which may correspond to JavaBean-style properties. An instance variable must be directly accessed only from within the methods of the entity by the entity instance itself. The state of the entity is available to clients only through the entity’s accessor methods (getter/setter methods) or other business methods.

`The state of the entity is available to clients only through the entity’s accessor methods (getter/setter methods) or other business methods.`

해당 내용을 보면 자바빈 규약에 맞춰서 꼭 작성하라는 얘기라기 보다는 필드에 직접 접근하는 것이 아닌
캡슐화를 권장하기 위해서 써있는 내용으로 보입니다. :)

phs1116 commented on src/main/java/techcourse/fakebook/domain/friendship/Friendship.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
equals, hashcode 다른 엔티티와 맞춰주세요. :)

phs1116 commented on src/main/java/techcourse/fakebook/domain/user/User.java:
```diff
@@ -89,6 +90,26 @@ public boolean checkEncryptedPassword(String encryptedPassword) {
     }
 
+    @Override
+    public boolean equals(Object o) {
```
엔티티의  equals, hascode는 ID로 식별합니다. 참고하세요.

eunsukko replied:
감사합니다..!

eunsukko replied:
id로 하려다가 질문이 생겼습니다..! (여기서 id는 `@Id`로 생성된 필드입니다.)

id의 경우 영속성 컨텍스트에 넣기전(null) 후(db에 적용된 id)가 달라집니다.
그런 경우.. 특히 Collection등을 사용할 때 문제가 생길수도 있다고 하구요 (ex. Set 에 넣은 후 id가 변경될 경우) 

또한 hibernate 문서에서는 `@NaturalId`를 추천해주던데
User의 경우 email이 이 역할을 할 수 있을 것으론 보였습니다. (비지니스 로직에서 id역할을 만족하는)
그런데 이 경우... 다른 엔티티들과의 통일성을 깨뜨리는 것 같다는 생각도 들고요.
그리고 혹시나... 비지니스 로직에서 id역할을 감당하지 못하게 될 수도 있고요.

------------------------
엔터티의 equals, hashcode 를 id로 모든 경우에 놓아도 괜찮을지 (Collection 등에서 생길법한 문제가 안생기도록 짜면서)
아니면 혹시 id로 놓지 않는 경우가 있다면 어떤 경우에 (혹은 어떤 기준으로) 그런 결정을 하시는지가 궁금합니다... :)

phs1116 replied:
데이터베이스 엔티티는 엔티티의 PK를 통해 식별 가능해야하고, 
아직 영속성에 등록되어있지 않은 객체는 데이터가 생성이 되지 않은 객체이기 때문에
말씀하신 부분에 대해서는 개발자가 주의해서 개발해야하는 수 밖에 없는 것 같습니다.

equals, hashcode에서 id만을 사용하지 않는 경우는 파티셔닝을 위해서 설정한것 외에는 보지 못한 것 같네요. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -1,45 +1,68 @@
+    private static final String DEFAULT_PATH = "src/main/resources/static";
+    private static final String ARTICLE_STATIC_FILE_PATH = "file/article/";
+
     private ArticleRepository articleRepository;
```
final을 선언해서 non null을 보장해줘도 좋을 것 같습니다. :)

DEFAULT_PATH, ARTICLE_STATIC_FILE_PATH 등 파일에 대한 설정은 이 서비스가 들고있는게 맞는지 다시 한번 고민해보시면 좋을 것 같네요.

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -1,45 +1,68 @@
+    private ArticleMultipartRepository articleMultipartRepository;
 
-    public ArticleService(ArticleRepository articleRepository, ArticleLikeRepository articleLikeRepository, UserService userService, ArticleAssembler articleAssembler) {
+    public ArticleService(ArticleRepository articleRepository, ArticleLikeRepository articleLikeRepository, UserService userService, ArticleAssembler articleAssembler, ArticleMultipartRepository articleMultipartRepository) {
```
의존하는 객체가 이제 5개가 넘어가는데요.
이제 ArticleService가 너무 많은 책임을 가지고 있지 않은지
도메인 서비스를 분리할 수 있을지 고민해봐도 좋을 것 같습니다.

vsh123 replied:
해당 내용에 대해서 팀원들과 상의를 해본 결과 시간적으로 오늘 안으로 반영하여 적용하기가 힘들 것 같다는 결론이 나왔습니다(서비스 분리의 필요성에 대해서는 확실히 인지가 되었습니다!)
 최대한 빨리 리팩토링 진행해서 다음주 리뷰 요청 때 반영할 수 있도록 하겠습니다!! :)
항상 감사합니다~~

phs1116 commented on src/main/java/techcourse/fakebook/domain/article/ArticleMultipart.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Objects;
+
+@Entity
+public class ArticleMultipart extends BaseEntity {
```
Multipart 라는 이름 자체는 파일을 전송하기 위한 프로토콜의 이름이니 엔티티 명으로는 적합하지 않은 것 같습니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -55,34 +78,67 @@ public void deleteById(Long id, UserOutline userOutline) {
-    private void checkAuthor(UserOutline userOutline, Article article) {
-        if (!article.getUser().isSameWith(userOutline.getId())) {
-            throw new InvalidAuthorException();
+    private Path writeFile(MultipartFile file, String hashingName) throws IOException {
```
ArticleService 내에 파일을 저장하는 책임을 상당부분 들고있는데요.
ArticleService의 책임이 맞는지 해당 역할을 담당하는 클래스를 만들수 없는지 고민해보시면 좋을 것 같습니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/CommentService.java:
```diff
@@ -70,33 +68,39 @@ public void deleteById(Long id, UserOutline userOutline) {
+    }
+
+    private Comment getComment(Long id) {
+        Comment comment = commentRepository.findById(id).orElseThrow(NotFoundCommentException::new);
```
findById시에 findByIdDeletedFalse로 JPA에 명시하거나
엔티티 클래스 어노테이션으로 '@Where(clause = "deleted = false")로 명시하면 조회시에 항상 삭제된 엔티티는 들고오지 않습니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/CommentService.java:
```diff
@@ -70,33 +68,39 @@ public void deleteById(Long id, UserOutline userOutline) {
+    }
+
+    public Integer getCommentsCountOf(Long articleId) {
+        return commentRepository.countCommentByArticleId(articleId);
```
삭제된 코멘트도 카운트 될 것 같은데 괜찮은가요?

phs1116 commented on src/main/java/techcourse/fakebook/service/FriendshipService.java:
```diff
@@ -0,0 +1,46 @@
+                .collect(Collectors.toList());
+    }
+
+    public void makeThemFriends(Long userId1, Long userId2) {
```
친구는 요청하는사람과 받는사람이 존재할텐데 user1, ueser2라는 변수명은 해당 모델이 드러나보이지 않아보이네요. :)

eunsukko replied:
현재 요청을 보내고 있지 않아서 (그냥... 친구초대하면.. 강제로 친구가 됩니다... ㅋㅋㅋㅋ) user1, user2 의 차이가 없다고 생각해서 이렇게 해놓았습니다..!

일단 바꿔보겠습니다..!!

이런 식으로 두 파라미터가 구분할 이유가 없을때 (예를 들어서 set에 2개를 추가한다고 했을때) 사용할 좋은 네이밍이 있을까요??....ㅎ

phs1116 replied:
현재 기능 상에서도 구분할 이유는 충분해 보여요. 
지금은 요청, 수락에 대한 프로세스는 없지만 해당 API을 요청해서 서비스를 호출한 당사자가 되는 유저와 
친구가 되는 유저는 존재하니까요. :) 

진짜 의미가 없다면 1,2 혹은 user,otherUser 등으로 표현할 수 있겠네요. :)

eunsukko replied:
감사합니다..! 
내부 구현 입장에서 생각해서 당사자와 대상을 구분할 필요가 없다고 생각한 것 같네요 ... ㅋㅋㅋㅋㅋㅋ

phs1116 commented on src/main/java/techcourse/fakebook/service/TotalService.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.stream.Collectors;
+
+@Service
+public class TotalService {
```
여러 도메인에 대한 내용을 facade 해주는 역할로 보이네요.
레이어 관점에서 보면  controller - {TotalService} - service 인 것 같은데요.
이 단계를 Application Layer라고 부르기도 합니다.

100% 참고하지 않아도 되지만 아래 블로그랑 관련된 내용 검색해보면서 공부해봐도 좋을 것 같아요.

https://brunch.co.kr/@cg4jins/7

phs1116 commented on src/main/java/techcourse/fakebook/service/dto/UserResponse.java:
```diff
@@ -1,14 +1,16 @@
+    private Long id;
     private String email;
     private String name;
     private String gender;
```
이전 단계에도 피드백 드렸던 내용인데요.
gender에 저장되는 내용 (FEMALE, MALE 같은)은 오로지 프론트에서 주는 데이터에만 의존되는 내용인가요?
이에 대한 별다른 타입 체크도 없이 그냥 저장하고 있는 것 같아서요.

어떤 값이 남성이고 어떤 값이 여성인지 등에 대해서 도메인에 전혀 보이지 않은 것 같아요.

eunsukko replied:
DB에 어떻게 넣을까 고민하다가 적용을 바로하지 못했었습니다. (birth, gender 같은 경우에요)
친구 관련기능 수정 후 적용하도록 하겠습니다..!

phs1116 commented on src/main/java/techcourse/fakebook/service/utils/ArticleAssembler.java:
```diff
@@ -18,6 +24,20 @@ public Article toEntity(ArticleRequest articleRequest, User user) {
+        return new TotalArticleResponse(articleResponse, countOfComment, countOfLike);
+    }
+
+    private LocalDateTime getRecentDate(Article article) {
```
해당 메서드는 Assembler가 가질 책임이 아니라 Article의 책임 아닐까요?

phs1116 commented on src/test/java/techcourse/fakebook/controller/CommentApiControllerTest.java:
```diff
@@ -75,61 +79,102 @@ void setUp() {
+        when().
+                get("/api/articles/" + articleResponse.getId() + "/comments/count").
+        then().
+                statusCode(200).
```
단순 숫자가 아니라 HttpStatus에 정의된 값으로 무슨 값인지도 같이 표현해주면 좋을 것 같네요. :)

phs1116 commented on src/main/java/techcourse/fakebook/domain/article/ArticleMultipart.java:
```diff
@@ -0,0 +1,64 @@
+
+    @ManyToOne
+    @OnDelete(action = OnDeleteAction.CASCADE)
+    private Article article;
```
현재 파일에 대해서는 Article에 대해서 단방향으로만 적용되어있는데요.
Article 조회시 항상 같이 조회되는 내용이라면 Article에서도 연관을 같이 가져가도 될 것 같습니다. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/dto/ArticleResponse.java:
```diff
@@ -1,16 +1,35 @@
     private String content;
+    private LocalDateTime recentDate;
     private UserOutline userOutline;
+    private List<ArticleMultipart> resources;
```
다른 엔티티들은 각자의 DTO로 돌려주는데 ArticleMultipart는 엔티티를 직접 돌려주네요.
통일성 있게 맞추는게 좋을 것 같네요.

phs1116 commented on src/main/java/techcourse/fakebook/service/ArticleService.java:
```diff
@@ -55,34 +78,67 @@ public void deleteById(Long id, UserOutline userOutline) {
+    public boolean like(Long id, UserOutline userOutline) {
+        Optional<ArticleLike> articleLike = Optional.ofNullable(articleLikeRepository.findByUserIdAndArticleId(userOutline.getId(), id));
+        if (articleLike.isPresent()) {
+            articleLikeRepository.delete(articleLike.get());
```
지금 어떤 엔티티는 delete = false 형태로 soft delete를 하고 있고요,
이 엔티티는 hard delete를 하고있네요.

삭제시에 어떻게 할지 전체적으로 정책을 통일해서 진행하는게 좋을 것 같습니다. :)

vsh123 replied:
위 내용에 팀원들과 회의를 해본 결과 ```엔티티의 보관 필요성(?)에 따라 soft를 할지 hard를 할지 결정을 하자```라는 결론이 나왔습니다. 
이에 따라 Like Entity는 hard,  Article,Comment는 Soft로 진행하고 추후 생성되는 엔티티에도 위와 같이 결정을 진행하겠습니다!

phs1116 replied:
네. 현재는 그렇게 해도 무방하지만 
실무에서는 둘 중 하나의 정책만 사용 가능하다는것도 참고해주세요. :)

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
전체적으로 구현 잘 해주셨어요. 👍 
테스트도 전반적으로 작성 잘 해주신 것 같아요.
피드백 몇가지 추가했는데 참고 부탁드릴게요. :)

wmakerjun commented on src/main/resources/templates/index.html:
```diff
@@ -29,8 +29,8 @@
+        <link rel="stylesheet" href="/css/index.css">
     </head>
     <body>
         <div id="wootecobook-app">
```
이전 작업한 부분에 바로 코멘트를 남길 수 없어 이곳에 남깁니다
https://github.com/woowacourse/miniprojects-2019/blob/94a8a65f9bd541308e4d331c515adf222ee4c970/src/main/resources/templates/index.html#L123-L242

현재 option에 직접적으로 모든 년도를 표시하고 있는데요. 아무래도 중복이 많기 때문에 자바스크립트를 이용해서 줄일 수 있다면 좋을 것 같아요!

[링크](https://stackoverflow.com/questions/23921850/how-to-make-the-selectbox-dropdown-short)를 참조해서 자바스크립트로 중복을 제거해 봅시다~!

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -1,6 +1,40 @@
+    }
+  }
+
   const ARTICLE_TEMPLATE_HTML =
```
템플릿 부분은 따로 파일을 분리해서 관리하면 좋을 것 같아요~!
그리고 html코드가 이렇게 많은 경우에는 최신자바스크립트 기법인 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

codeanddonuts replied:
혹시 분리하려면 ajax로 불러오는 방법 말고 없나요?

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -1,6 +1,40 @@
+      return e.keyCode === ENTER || e.which === ENTER || e.key === ENTER
+    }
+
+    formatDate(dateString) {
```
우와 시간표시를 하기 위해 꼼꼼하게 구현한 점 너무 멋집니다 💯 
혹시 시간과 관련된 표시를 위해서 다른 기능을 추가하게 된다면 [momentjs](https://momentjs.com) 라이브러리를 활용하면 더욱 쉽고 빠르게 구현할 수도 있어요~! 참고해주길 바래요 :)

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -35,235 +70,204 @@ const App = (() => {
     }
 
-    const remove = async id => {
+    async remove(id) {
```
비동기 통신하는 부분에 async await을 사용하려한 점이 인상적이에요~!
다만 [axios](https://github.com/axios/axios) 공식 예제들을 보면 알겠지만, 지금 구현한 코드의 로직을 공식예제에 있는 axios.method().then().catch() 형태로 모두 구현할 수 있는 로직들인 것 같아요. async await은 비동기 통신에서 순서를 보장하기 위해서 사용하는 경우가 많은데 지금 구현한 로직은 반드시 async await를 쓰지 않아도 될 것 같아 보여서 추가적인 코멘트 남깁니다 :)

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -35,235 +70,204 @@ const App = (() => {
+          if(files.length > 0) {
+            req.append("files", files[0])
+          }
+          const article = (await axios.post(BASE_URL + "/api/articles", req)).data
```
👍 axios 라이브러리를 이용해서 ajax 통신하다니 멋져요. 여기서 한 단계 더 나아가기 위해서 api만 관리하는 객체를 하나 더 만들면 어떨까요? api를 아래처럼 관리하다면 역할을 더 명확히 구분할 수 있고, api 관리가 더욱 수월해질거에요 😆 

```javascript

const Api = function () {
    const domain = 'http://localhost:8080'

    const request = {
      get(path) {
        return axios.get(`${domain + path}`)
      },
      post(path, data) {
        return axios.post(`${domain + path}`, data)
      },
      delete(path) {
        return axios.delete(`${domain + path}`)
      },
      put(path, data) {
        return axios.put(`${domain + path}`, data)
      }
    }

    const card = {
      fetch(id) {
        return request.get(`/cards/${id}`)
      },
      create({title, listId, pos}) {
        return request.post(`/cards`, {title, listId, pos})
      },
      update(id, data) {
        return request.put(`/cards/${id}`, data)
      },
      destroy(id) {
        return request.delete(`/cards/${id}`)
      }
    }

    return {
        card: card
    }
}
```

wmakerjun commented on src/main/resources/templates/newsfeed.html:
```diff
@@ -175,114 +201,108 @@
+            <div class="mrg-top-20"> </div>
         </footer>
+        <script type="text/javascript">
+        window.onload = function(){
```
window.onload를 여기에 꼭 써야 하는 이유가 있을까요?
body 태그가 닫히기 전에 자바스크립가 load되므로 굳이 window.onload를 쓸 필요가 없어보여요. window.onload는 성능을 저하 시킬 수 있기 때문에 주의해서 사용해야 합니다.
아래 링크를 참고하면 좋을 것 같습니다 :)
http://jinolog.com/programming/javascript/2011/03/21/jquery-ready-function.html

men7627 replied:
반영하였습니다!
해당 함수 선언과 동시에 끝에 '()'를 붙여줘서 호출까지 할 수 있도록 하였습니다!

choihz committed:
refactor: 프론트 ui 수정 (#74)

* refactor: 글쓰기 관련 버튼 마우스 오버 효과 및 링크 범위 재설정

* refactor: 댓글 간격 수정 및 불필요한 선 제거

* refactor: 게시글 하단 및 footer 여백 추가

* refactor: profile 페이지와 newsfeed 페이지 폼 디자인 통일, 주석 제거

* feat: 프로필 페이지 친구 이름 마크업

* feat: 댓글 개수 조회 api 구현

* refactor: 글쓰기 좋아요 디자인 개선

wmakerjun commented on src/main/resources/templates/newsfeed.html:
```diff
@@ -175,114 +201,108 @@
+         }
+
+         function selectData(that) {
+             var search = document.getElementById("search");
```
var 대신에 let이나 const를 이용해주세요 :)
https://www.daleseo.com/js-var-issues/

men7627 replied:
반영하였습니다~!

wmakerjun commented on src/main/resources/templates/newsfeed.html:
```diff
@@ -175,114 +201,108 @@
+        window.onload = function(){
+            const SPACE_PATTERN = /\s/g;
+
+            let search = document.getElementById("search");
```
변수 자체를 재할당하는 것이 아니면은 let 보다는 const를 이용하는게 더 좋습니다 :)
https://smilerici.tistory.com/68

wmakerjun reviewed (changes requested):
프론트엔드 구현하느라 너무 고생하셨습니다~! 코드에서 반복적으로 등장하는 실수들 위주로 코멘트 남겼으니, 피드백 반영하면 훨씬 더 많이 성장할 것 같아요 😃화이팅입니다!!! 😆

choihz committed:
refactor: 컨트롤러 리팩토링 (#76)

* fix: ResponseEntity의 바디가 없는 경우 제네릭을 Void로 수정

* refactor: HttpStatus 사용하여 테스트 가독성 향상

* fix: ResponseEntity에 제공하는 URI를 적절하게 수정

* refactor: Controller, Service의 필드에 final 키워드 추가

men7627 committed:
refactor: 검색기능 css, js 분리 (#80)

* fix: 회원가입 후 자동로그인 구현

1. POST("/users")에 login로직, session set로직을 추가한다.
2. signup후 login하는 테스트를 signup만 하게 수정한다.

* feat: 자동완성 유저검색 기능 구현

1. ajax 통신을 활용해 검색창 datalist를 동적으로 변하게 구현한다.
 - 한글자 입력할 때마다 api요청
2. GET("api/users/{keyword}")를 구현한다.
3. UserRepository에 findByNameContaining(String) 메서드를 추가한다
4. UserService에 findUserNamesByKeyword(String) 메서드를 추가한다.
5. 각 테스트를 추가한다.

* feat: 유저이름 조회 테스트 간소화

* feat: 유저이름 검색 스크롤 추가

* refactor: 검색 css, js 분리

vsh123 committed:
refactor: 파일업로드, 게시글, 댓글 (#81)

* refactor: Service의 의존 non null 보장을 위해 final 설정

* rename: ArticleMultipart -> ArticleAttachment

* refactor: 파일 업로드 서비스 분리, ArticleAttachment에 대한 dto 생성

* refactor: Article, Comment @where 어노테이션 추가

* refactor: Article ArtcileAttachment 연관관계 매핑

vsh123 committed:
feat: nesfeed 로딩 시 이미지 추가 (#84)

choihz committed:
feat: 프로필 페이지 구현 (#89)

* feat: 프로필 페이지 구현

* fix: 테스트 오류 수정

* feat: 프로필 페이지 로딩 시 이미지 추가

* fix: 경로 오류 수정

vsh123 committed:
feat: 검색 기능 추가 구현 (#90)

* fix: 회원가입 후 자동로그인 구현

1. POST("/users")에 login로직, session set로직을 추가한다.
2. signup후 login하는 테스트를 signup만 하게 수정한다.

* feat: 자동완성 유저검색 기능 구현

1. ajax 통신을 활용해 검색창 datalist를 동적으로 변하게 구현한다.
 - 한글자 입력할 때마다 api요청
2. GET("api/users/{keyword}")를 구현한다.
3. UserRepository에 findByNameContaining(String) 메서드를 추가한다
4. UserService에 findUserNamesByKeyword(String) 메서드를 추가한다.
5. 각 테스트를 추가한다.

* feat: 유저이름 조회 테스트 간소화

* feat: 유저이름 검색 스크롤 추가

* refactor: 검색 css, js 분리

* fix: 검색 리스트 이메일 추가 & 마이페이지 검색 추가

1. 동명이인을 고려해 검색 리스트에 이메일을 같이 추가한다.
2. 검색 버튼을 누르면 해당 유저의 페이지로 이동한다.
3. 마이페이지에서도 검색을 가능하게 한다.

woowahan-pjs commented on src/test/java/techcourse/fakebook/controller/FriendshipApiControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+    private List<Long> userIds;
+
+
+    private Long loginedUserId;
```
'logined'보다 'logged-in'가 더 정확해 보여요.

eunsukko replied:
감사합니다..!

혹시 클래스명의 경우 '-' 가 들어가면 안되는 것 같은데.. 
'-' 를 넣으면 클래스명과 변수명의 통일성을 해치진 않을까요???

woowahan-pjs replied:
여기서 'logged-in'는 영문 표기입니다. 클래스 이름 혹은 변수 이름으로 사용될 경우 `loggedIn`이겠죠?

woowahan-pjs commented on src/main/java/techcourse/fakebook/controller/utils/ExceptionAdvice.java:
```diff
@@ -0,0 +1,11 @@
+package techcourse.fakebook.controller.utils;
```
`controller.utils`? 어떤 패키지에 두는 것이 좋을지 다른 스프링 웹 애플리케이션의 사례를 찾아보세요.

woowahan-pjs commented on src/main/java/techcourse/fakebook/controller/UserWebController.java:
```diff
@@ -6,10 +6,15 @@
+import techcourse.fakebook.service.dto.LoginRequest;
+import techcourse.fakebook.service.dto.UserOutline;
 import techcourse.fakebook.service.dto.UserResponse;
 import techcourse.fakebook.service.dto.UserSignupRequest;
```
'signup'보다 'sign-up'이 더 정확해 보여요.

woowahan-pjs commented on src/main/java/techcourse/fakebook/Fakebook.java:
```diff
@@ -10,4 +10,4 @@
     public static void main(String[] args) {
         SpringApplication.run(Fakebook.class, args);
```
메인 클래스 이름은 `FakebookApplication` 혹은 `Application`이라고 쓰는 것이 관례입니다.

woowahan-pjs reviewed (changes requested):
VALUE OBJECT로 대신할 수 있는 DTO가 보입니다.
도메인 패키지에서 의미가 있는 클래스를 더 도출해서 성숙한 소프트웨어를 만들어 보세요.
객체 지향 생활 체조 원칙을 최대한 적용해 보세요.

men7627 committed:
refactor: 검색 창 디자인 수정 (#91)

1. 검색 리스트 창이 동적으로 바뀐다.
2. 180px를 넘으면 스크롤이 생긴다.

choihz committed:
feat: 댓글 출력 구현 (#95)

eunsukko committed:
feat: 친구추가, 제거 기능 (+ 뉴스피드에 친구글도 보이도록) (#93)

* feat: logger 에서 매소드명 출력하도록 수정

* fix: id를 통한 equals, hashcode 구현 (User, Friendship 엔터티)

* feat: 친구추가 적용(프로필에 있는 버튼을 통한)

* feat: 친구 제거기능 추가

* feat: 뉴스피드에 친구게시글도 추가

* refactor: 친구제거 함수명 변경

* fix: 비밀번호에서 사용가능한 특수문자 수정

* fix: 친구추가,제거시 잘못된 경우 BadRequest(400) 전송

choihz committed:
fix: 전체적인 오류 수정 (#99)

* fix: 프로필 페이지 상단에 로그인한 유저 정보가 보이도록 수정

* fix: 검색창 스크롤바가 계속 보이는 오류 수정

* fix: TotalService 오류 수정

* fix: 깨지는 테스트 수정

* fix: 게시글 및 댓글을 글 작성일 기준으로 적절히 정렬하여 가져오도록 수정

* refactor: 메인 클래스 이름 수정

codeanddonuts committed:
refactor: front-end scripts (#98)

* refactor: front-end scripts

* fix: article editing

eunsukko committed:
feat: 친구추가/제거 프론트에 버튼 적용 (#100)
